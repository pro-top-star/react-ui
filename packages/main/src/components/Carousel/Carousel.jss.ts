import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

const styles = {
  carousel: {
    position: 'relative',
    overflow: 'hidden',
    boxSizing: 'border-box',
    border: '1px solid transparent',
    touchAction: 'pan-y',
    minWidth: '15.5rem',
    fontFamily: ThemingParameters.sapFontFamily,
    backgroundColor: ThemingParameters.sapBackgroundColor,
    '&:focus': {
      outline: 'none',
      '&:before': {
        border: '1px solid #000000',
        position: 'absolute',
        content: '" "',
        top: '0',
        right: '0',
        bottom: '0',
        left: '0',
        zIndex: '2',
        pointerEvents: 'none'
      },
      '&:after': {
        border: '1px dotted #ffffff',
        position: 'absolute',
        content: '" "',
        top: '0',
        right: '0',
        bottom: '0',
        left: '0',
        zIndex: '2',
        pointerEvents: 'none'
      }
    }
  },
  carouselInner: {
    height: '100%',
    position: 'relative',
    whiteSpace: 'nowrap',
    textAlign: 'start',
    fontSize: '0',
    transition: 'transform 0.5s cubic-bezier(0.46, 0, 0.44, 1)'
  },
  carouselItem: {
    width: '100%',
    height: '100%',
    overflow: 'hidden',
    display: 'inline-block',
    verticalAlign: 'top',
    whiteSpace: 'normal',
    fontSize: '1rem',
    visibility: 'hidden'
  },
  carouselArrowPlacementContent: {
    '&:hover': {
      '& [data-value="paginationArrow"]': {
        opacity: 1,
        '& ui5-icon': {
          transform: 'rotate(0deg)'
        }
      }
    },
    '& $carouselItem': {
      padding: '0 4rem',
      width: 'calc(100% - 8rem)'
    }
  }
};

export default styles;
