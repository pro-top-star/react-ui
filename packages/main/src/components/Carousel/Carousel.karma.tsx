import { getEventFromCallback, mountThemedComponent } from '@shared/tests/utils';
import { expect, use } from 'chai';
import { matchSnapshot } from 'chai-karma-snapshot';
import React, { cloneElement } from 'react';
import * as sinon from 'sinon';
import { Carousel } from '../../lib/Carousel';
import { CarouselArrowsPlacement } from '../../lib/CarouselArrowsPlacement';
import { Icon } from '../../lib/Icon';
import { PlacementType } from '../../lib/PlacementType';

use(matchSnapshot);

const renderCarousel = (props) => {
  return (
    <Carousel {...props}>
      <p>Carousel 1</p>
      <p>Carousel 2</p>
      <p>Carousel 3</p>
      <p>Carousel 4</p>
      <p>Carousel 5</p>
      <p>Carousel 6</p>
      <p>Carousel 7</p>
    </Carousel>
  );
};

describe('Carousel', () => {
  it('Renders without crashing', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 0 }));
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('Should render a text indicator', () => {
    const wrapper = mountThemedComponent(
      <Carousel>
        <p>Carousel 1</p>
        <p>Carousel 2</p>
        <p>Carousel 3</p>
        <p>Carousel 4</p>
        <p>Carousel 5</p>
        <p>Carousel 6</p>
        <p>Carousel 7</p>
        <p>Carousel 8</p>
        <p>Carousel 9</p>
        <p>Carousel 10</p>
      </Carousel>
    );
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('CarouselArrowsPlacement: Content', () => {
    const wrapper = mountThemedComponent(renderCarousel({ arrowsPlacement: CarouselArrowsPlacement.Content }));
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('CarouselArrowsPlacement: PageIndicator', () => {
    const wrapper = mountThemedComponent(renderCarousel({ arrowsPlacement: CarouselArrowsPlacement.PageIndicator }));
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('Page Indicator Placement: Top', () => {
    const wrapper = mountThemedComponent(
      renderCarousel({
        arrowsPlacement: CarouselArrowsPlacement.PageIndicator,
        pageIndicatorPlacement: PlacementType.Top
      })
    );
    expect(wrapper.debug()).to.matchSnapshot();
    wrapper.setProps({
      children: cloneElement(wrapper.props().children, { activePage: 1 })
    });
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(1);
  });

  it('Update activePage via prop', () => {
    const wrapper = mountThemedComponent(
      renderCarousel({
        activePage: 0
      })
    );
    expect(wrapper.debug()).to.matchSnapshot();
  });

  it('Navigation to next page', () => {
    const callback = sinon.spy();
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 0, onPageChanged: callback }));
    wrapper
      .find(Icon)
      .last()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(1);
    expect(getEventFromCallback(callback).getParameter('selectedIndex')).to.equal(1);
  });

  it('Navigation to previous page', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 1 }));
    wrapper
      .find(Icon)
      .first()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(0);
  });

  it('Navigation to previous page - w/o Loop', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 0 }));
    wrapper
      .find(Icon)
      .first()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(0);
  });

  it('Navigation to previous page - w/ Loop', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 0, loop: true }));
    wrapper
      .find(Icon)
      .first()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(6);
  });

  it('Navigation to next page - w/o Loop', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 6 }));
    wrapper
      .find(Icon)
      .last()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(6);
  });

  it('Navigation to next page - w/ Loop', () => {
    const wrapper = mountThemedComponent(renderCarousel({ activePage: 6, loop: true }));
    wrapper
      .find(Icon)
      .last()
      .simulate('click');
    wrapper.update();
    // @ts-ignore
    const instance = wrapper.find(Carousel.InnerComponent).instance();
    // @ts-ignore
    expect(instance.state.activePage).to.equal(0);
  });

  it('Carousel with 1 child', () => {
    const wrapper = mountThemedComponent(
      <Carousel arrowsPlacement={CarouselArrowsPlacement.Content}>
        <p>Carousel 1</p>
      </Carousel>
    );
    expect(wrapper.render().find('[data-value="paginationArrow"]')).to.have.length(0);
  });
});
