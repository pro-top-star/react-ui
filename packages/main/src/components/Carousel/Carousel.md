```jsx
import { Carousel } from '@ui5/webcomponents-react/lib/Carousel';
```
