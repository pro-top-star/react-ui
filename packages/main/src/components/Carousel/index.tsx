import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { CarouselArrowsPlacement } from '@ui5/webcomponents-react/lib/CarouselArrowsPlacement';
import { PlacementType } from '@ui5/webcomponents-react/lib/PlacementType';
import React, {
  Children,
  CSSProperties,
  FC,
  forwardRef,
  ReactNode,
  ReactNodeArray,
  Ref,
  useCallback,
  useEffect,
  useState
} from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';
import styles from './Carousel.jss';
import { CarouselPagination, CarouselPaginationPropTypes } from './CarouselPagination';

export interface CarouselPropTypes
  extends Omit<CarouselPaginationPropTypes, 'goToPreviousPage' | 'goToNextPage'>,
    CommonProps {
  /**
   * The content which the carousel displays.
   */
  children: ReactNode | ReactNodeArray;
  /**
   * This event is fired after a carousel swipe has been completed
   */
  onPageChanged?: (event: Event) => void;
  /**
   * The height of the carousel. Note that when a percentage value is used, the height of the surrounding container
   * must be defined.
   */
  height?: CSSProperties['height'];
  /**
   * The width of the carousel. Note that when a percentage value is used, the height of the surrounding container
   * must be defined.
   */
  width?: CSSProperties['width'];
  /**
   * Defines whether the carousel should loop, i.e show the first page after the last page is reached and vice versa.
   */
  loop?: boolean;
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof styles>>(styles, { name: 'Carousel' });

const Carousel: FC<CarouselPropTypes> = forwardRef((props: CarouselPropTypes, ref: Ref<HTMLDivElement>) => {
  const {
    children,
    height,
    width,
    pageIndicatorPlacement,
    className,
    style,
    arrowsPlacement,
    tooltip,
    showPageIndicator,
    loop,
    onPageChanged
  } = props;

  const [currentlyActivePage, setCurrentlyActivePage] = useState(props.activePage);

  useEffect(() => {
    setCurrentlyActivePage(props.activePage);
  }, [props.activePage, setCurrentlyActivePage]);

  const outerStyle = { width, height };
  if (style) {
    Object.assign(outerStyle, style);
  }

  const classes = useStyles();
  const classNameString = StyleClassHelper.of(classes.carousel);

  if (className) {
    classNameString.put(className);
  }

  const carouselItemClasses = StyleClassHelper.of(classes.carouselItem);
  if (arrowsPlacement === CarouselArrowsPlacement.Content) {
    classNameString.put(classes.carouselArrowPlacementContent);
  }

  const selectPageAtIndex = useCallback(
    (index, event) => {
      setCurrentlyActivePage(index);
      onPageChanged(Event.of(null, event, { selectedIndex: index }));
    },
    [onPageChanged, setCurrentlyActivePage]
  );

  const childElementCount = Children.count(children);
  const goToNextPage = useCallback(
    (e, skipManualInversion = false) => {
      if (
        document.dir === 'rtl' &&
        arrowsPlacement === CarouselArrowsPlacement.Content &&
        e.type === 'click' &&
        !skipManualInversion
      ) {
        return goToPreviousPage(e, true);
      }
      if (loop === false && currentlyActivePage === childElementCount - 1) {
        return;
      }
      const nextPage = currentlyActivePage === childElementCount - 1 ? 0 : currentlyActivePage + 1;
      selectPageAtIndex(nextPage, e);
    },
    [loop, currentlyActivePage, selectPageAtIndex, childElementCount, arrowsPlacement]
  );

  const goToPreviousPage = useCallback(
    (e, skipManualInversion = false) => {
      if (
        document.dir === 'rtl' &&
        arrowsPlacement === CarouselArrowsPlacement.Content &&
        e.type === 'click' &&
        !skipManualInversion
      ) {
        return goToNextPage(e, true);
      }
      if (loop === false && currentlyActivePage === 0) {
        return;
      }
      const previousPage = currentlyActivePage === 0 ? childElementCount - 1 : currentlyActivePage - 1;
      selectPageAtIndex(previousPage, e);
    },
    [loop, childElementCount, currentlyActivePage, selectPageAtIndex, arrowsPlacement, goToNextPage]
  );

  const onKeyDown = useCallback(
    (e) => {
      if (e.key === 'ArrowRight') {
        if (document.dir === 'rtl') {
          goToPreviousPage(e);
        } else {
          goToNextPage(e);
        }
      }
      if (e.key === 'ArrowLeft') {
        if (document.dir === 'rtl') {
          goToNextPage(e);
        } else {
          goToPreviousPage(e);
        }
      }
    },
    [goToPreviousPage, goToNextPage]
  );

  const translateXPrefix = document.dir === 'rtl' ? '' : '-';

  const passThroughProps = usePassThroughHtmlProps(props);

  return (
    <div
      className={classNameString.toString()}
      style={outerStyle}
      title={tooltip}
      slot={props['slot']}
      ref={ref}
      role="list"
      tabIndex={0}
      onKeyDown={onKeyDown}
      {...passThroughProps}
    >
      {childElementCount > 1 && pageIndicatorPlacement === PlacementType.Top && (
        <CarouselPagination
          arrowsPlacement={arrowsPlacement}
          showPageIndicator={showPageIndicator}
          pageIndicatorPlacement={pageIndicatorPlacement}
          activePage={currentlyActivePage}
          children={children}
          goToPreviousPage={goToPreviousPage}
          goToNextPage={goToNextPage}
        />
      )}
      <div
        className={classes.carouselInner}
        style={{
          transform: `translateX(${translateXPrefix}${currentlyActivePage * 100}%)`
        }}
      >
        {Children.map(children, (item, index) => (
          <div
            key={index}
            className={carouselItemClasses.toString()}
            role="listitem"
            aria-posinset={index}
            aria-setsize={childElementCount}
            style={{ visibility: index === currentlyActivePage ? 'visible' : null }}
          >
            {item}
          </div>
        ))}
      </div>
      {childElementCount > 1 && pageIndicatorPlacement === PlacementType.Bottom && (
        <CarouselPagination
          arrowsPlacement={arrowsPlacement}
          showPageIndicator={showPageIndicator}
          pageIndicatorPlacement={pageIndicatorPlacement}
          activePage={currentlyActivePage}
          children={children}
          goToPreviousPage={goToPreviousPage}
          goToNextPage={goToNextPage}
        />
      )}
    </div>
  );
});

Carousel.displayName = 'Carousel';
Carousel.defaultProps = {
  activePage: 0,
  arrowsPlacement: CarouselArrowsPlacement.Content,
  onPageChanged: () => {},
  height: '100%',
  width: '100%',
  showPageIndicator: true,
  loop: false,
  pageIndicatorPlacement: PlacementType.Bottom
};

export { Carousel };
