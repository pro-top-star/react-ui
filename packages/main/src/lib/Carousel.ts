import { Carousel } from '../webComponents/Carousel';

export { Carousel };
