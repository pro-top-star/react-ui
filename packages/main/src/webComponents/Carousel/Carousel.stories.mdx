import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { Carousel } from "@ui5/webcomponents-react/lib/Carousel";
import { CarouselArrowsPlacement } from "@ui5/webcomponents-react/lib/CarouselArrowsPlacement";
import { createSelectArgTypes } from "@shared/stories/createSelectArgTypes";
import { DocsHeader } from "@shared/stories/DocsHeader";
import { DocsCommonProps } from "@shared/stories/DocsCommonProps";

<Meta
  title="UI5 Web Components / Carousel"
  component={Carousel}
  argTypes={{
    ...createSelectArgTypes({ arrowsPlacement: CarouselArrowsPlacement }),
    ...DocsCommonProps,
    children: { control: { disable: true } },
  }}
  args={{
    arrowsPlacement: CarouselArrowsPlacement.Content,
    style: {},
    className: "",
    tooltip: "",
    slot: "",
    ref: null,
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return (
        <Carousel {...args}>
          <img
            src="https://sap.github.io/ui5-webcomponents/assets/images/sample1.jpg"
            alt="img-sample 1"
          />
          <img
            src="https://sap.github.io/ui5-webcomponents/assets/images/sample2.jpg"
            alt="img-sample 2"
          />
          <img
            src="https://sap.github.io/ui5-webcomponents/assets/images/sample3.jpg"
            alt="img-sample 3"
          />
        </Carousel>
      );
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />
<div
  style={{
    fontFamily: "var(--sapFontFamily)",
    fontSize: "var(--sapFontSize)",
    color: "var(--sapTextColor)",
  }}
>
  <h3>Usage</h3>
  <b>When to use:</b> <br />
  The items you want to display are very different from each other. - You want
  to display the items one after the other. <br />
  <b>When not to use:</b>
  <br />
  The items you want to display need to be visible at the same time. - The items
  you want to display are uniform and very similar For the <code>Carousel</code>
</div>
