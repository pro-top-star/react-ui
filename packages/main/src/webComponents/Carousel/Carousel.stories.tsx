import { boolean, number, select } from '@storybook/addon-knobs';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import { Carousel } from '@ui5/webcomponents-react/lib/Carousel';
import { CarouselArrowsPlacement } from '@ui5/webcomponents-react/lib/CarouselArrowsPlacement';
import React from 'react';

export default {
  title: 'UI5 Web Components / Carousel',
  component: Carousel
};

export const generatedDefaultStory = () => (
  <Carousel
    arrowsPlacement={select('arrowsPlacement', CarouselArrowsPlacement, CarouselArrowsPlacement['Content'])}
    cycling={boolean('cycling', false)}
    hideNavigation={boolean('hideNavigation', false)}
    itemsPerPage={number('itemsPerPage', 1)}
    selectedIndex={number('selectedIndex', 0)}
  >
    <div style={{ height: '300px', width: '100%', backgroundColor: ThemingParameters.sapInformationBackground }}>
      Carousel Content 1
    </div>
    <div style={{ height: '300px', width: '100%', backgroundColor: ThemingParameters.sapInformationBackground }}>
      Carousel Content 2
    </div>
    <div style={{ height: '300px', width: '100%', backgroundColor: ThemingParameters.sapInformationBackground }}>
      Carousel Content 3
    </div>
    <div style={{ height: '300px', width: '100%', backgroundColor: ThemingParameters.sapInformationBackground }}>
      Carousel Content 4
    </div>
  </Carousel>
);

generatedDefaultStory.story = {
  name: 'Generated default story'
};
