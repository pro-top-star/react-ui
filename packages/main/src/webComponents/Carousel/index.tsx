import { CarouselArrowsPlacement } from '@ui5/webcomponents-react/lib/CarouselArrowsPlacement';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import UI5Carousel from '@ui5/webcomponents/dist/Carousel';
import React, { FC, ReactNode } from 'react';
import { WithWebComponentPropTypes } from '../../internal/withWebComponent';

export interface CarouselPropTypes extends WithWebComponentPropTypes {
  /**
   * Defines the position of arrows. <br><br> Available options are: <ul> <li><code>Content</code></li> <li><code>Navigation</code></li> </ul> <br> When set to "Content", the arrows are placed on the sides of the current page. <br> When set to "Navigation", the arrows are placed on the sides of the page indicator.
   */
  arrowsPlacement?: CarouselArrowsPlacement;
  /**
   * Defines whether the carousel should loop, i.e show the first page after the last page is reached and vice versa.
   */
  cycling?: boolean;
  /**
   * If set to true the navigation is hidden.
   */
  hideNavigation?: boolean;
  /**
   * Sets the amount of items per page. If this property is set, on mobile devices it will always fallback to 1.
   */
  itemsPerPage?: number;
  /**
   * Defines the index of the initially selected page.
   */
  selectedIndex?: number;
  /**
   * Defines the content of the <code>ui5-carousel</code>.
   */
  children?: ReactNode | ReactNode[];
}

/**
 * <code>import { Carousel } from '@ui5/webcomponents-react/lib/Carousel';</code>
 * <br />
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/Carousel" target="_blank">UI5 Web Components Playground</a>
 */
const Carousel: FC<CarouselPropTypes> = withWebComponent<CarouselPropTypes>(UI5Carousel);

Carousel.displayName = 'Carousel';

Carousel.defaultProps = {
  arrowsPlacement: CarouselArrowsPlacement.Content,
  cycling: false,
  hideNavigation: false,
  itemsPerPage: 1,
  selectedIndex: 0
};

export { Carousel };
