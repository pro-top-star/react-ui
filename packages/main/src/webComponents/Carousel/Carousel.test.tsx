import { Carousel } from '@ui5/webcomponents-react/lib/Carousel';
import { mount } from 'enzyme';
import React from 'react';

describe('Carousel', () => {
  test('Basic Test (generated)', () => {
    const wrapper = mount(<Carousel />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
