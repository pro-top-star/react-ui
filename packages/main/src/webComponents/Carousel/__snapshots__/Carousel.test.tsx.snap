// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Basic Test (generated) 1`] = `
<ui5-carousel
  arrows-placement="Content"
  items-per-page="1"
  selected-index="0"
/>
`;
